pkg/kubelet/cm/dra/manager.go

```diff
@@ -0,0 +1,255 @@
+type ManagerImpl struct {
+	// cache contains cached claim info
+	cache *claimInfoCache
+
+	// KubeClient reference
+	kubeClient clientset.Interface
+}
+
+// NewManagerImpl creates a new manager.
+func NewManagerImpl(kubeClient clientset.Interface) (*ManagerImpl, error) {
+	klog.V(2).InfoS("Creating DRA manager")
+
+	manager := &ManagerImpl{
+		cache:      newClaimInfoCache(),
+		kubeClient: kubeClient,
+	}
+
+	return manager, nil
+}
+
+// Generate container annotations using CDI UpdateAnnotations API.
+func generateCDIAnnotations(
+	claimUID types.UID,
+	driverName string,
+	cdiDevices []string,
+) ([]kubecontainer.Annotation, error) {
+	annotations, err := updateAnnotations(map[string]string{}, driverName, string(claimUID), cdiDevices)
+	if err != nil {
+		return nil, fmt.Errorf("can't generate CDI annotations: %+v", err)
+	}
+
+	kubeAnnotations := []kubecontainer.Annotation{}
+	for key, value := range annotations {
+		kubeAnnotations = append(kubeAnnotations, kubecontainer.Annotation{Name: key, Value: value})
+	}
+
+	return kubeAnnotations, nil
+}
+
+// prepareContainerResources attempts to prepare all of required resource
+// plugin resources for the input container, issue an NodePrepareResource rpc request
+// for each new resource requirement, process their responses and update the cached
+// containerResources on success.
+func (m *ManagerImpl) prepareContainerResources(pod *v1.Pod, container *v1.Container) error {
+	// Process resources for each resource claim referenced by container
+	for range container.Resources.Claims {
+		for i, podResourceClaim := range pod.Spec.ResourceClaims {
+			claimName := resourceclaim.Name(pod, &pod.Spec.ResourceClaims[i])
+			klog.V(3).InfoS("Processing resource", "claim", claimName, "pod", pod.Name)
+
+			if claimInfo := m.cache.get(claimName, pod.Namespace); claimInfo != nil {
+				// resource is already prepared, add pod UID to it
+				claimInfo.addPodReference(pod.UID)
+
+				continue
+			}
+
+			// Query claim object from the API server
+			resourceClaim, err := m.kubeClient.ResourceV1alpha1().ResourceClaims(pod.Namespace).Get(
+				context.TODO(),
+				claimName,
+				metav1.GetOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to fetch ResourceClaim %s referenced by pod %s: %+v", claimName, pod.Name, err)
+			}
+
+			// Check if pod is in the ReservedFor for the claim
+			if !resourceclaim.IsReservedForPod(pod, resourceClaim) {
+				return fmt.Errorf("pod %s(%s) is not allowed to use resource claim %s(%s)",
+					pod.Name, pod.UID, podResourceClaim.Name, resourceClaim.UID)
+			}
+
+			// Call NodePrepareResource RPC
+			driverName := resourceClaim.Status.DriverName
+
+			client, err := dra.NewDRAPluginClient(driverName)
+			if err != nil {
+				return fmt.Errorf("failed to get DRA Plugin client for plugin name %s, err=%+v", driverName, err)
+			}
+
+			response, err := client.NodePrepareResource(
+				context.Background(),
```

Comment:
One more thing I think we should note for beta: I don't see any timeout / context associated with GRPC call to the plugin. Since this is in the container start path, I suppose it means we can block container start if the plugin is not responding/not up. We should consider if we need to add a timeout here. (Worth to create a tracking issue?)

Reply:
Yes, definitely worth an issue. I can already see that our tracking board will be pretty full once I get around to triaging all the issues filed this week - good!

=> https://github.com/kubernetes/kubernetes/issues/113867